[Style] Convert the 'glyph-orientation-*' properties to strong style types

weinig opened the pull request:
#### 61c7aa15477cdcc0925925cc26e328419fdc9f1e
<pre>
[Style] Convert the &apos;glyph-orientation-*&apos; properties to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=300064">https://bugs.webkit.org/show_bug.cgi?id=300064</a>

Reviewed by Darin Adler.

Converts the &apos;glyph-orientation-horizontal&apos; and &apos;glyph-orientation-vertical&apos;
properties to strong style types.

To make properties like these, ones that can represented purely as `enum`,
work with strong style types, a new interface, `ValueRepresentation`, has
been added that allows converting the enum to its counterpart. This allows
us to implement a single interface and getting CSSValueCreation, Serialization
and Logging all for free. I expect this interface to evolve as we convert
more `enum` properties (for example, should we have a bidirectional mapping
of enum values &lt;-&gt; CSSValueIDs for simpler conversions, probably), but this
gives us a good starting out point.

To make it clear which enums still need `ValueRepresentation` conformance,
enums that don&apos;t have it (and are used with the style system) will cause
a compile error unless they are added to the list in StylePrimitiveKeyword+ValueRepresentationNeeded.h

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/ViewTimeline.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/CSSValueConcepts.h:
* Source/WebCore/page/IntersectionObserver.cpp:
* Source/WebCore/rendering/RenderTreeAsText.cpp:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/OutlineValue.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.cpp:
* Source/WebCore/rendering/style/StyleAppleColorFilterData.cpp:
* Source/WebCore/rendering/style/StyleBoxData.cpp:
* Source/WebCore/rendering/style/StyleDeprecatedFlexibleBoxData.cpp:
* Source/WebCore/rendering/style/StyleFilterData.cpp:
* Source/WebCore/rendering/style/StyleFlexibleBoxData.cpp:
* Source/WebCore/rendering/style/StyleFontData.cpp:
* Source/WebCore/rendering/style/StyleGridData.cpp:
* Source/WebCore/rendering/style/StyleInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMarqueeData.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMultiColData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleSurroundData.cpp:
* Source/WebCore/rendering/style/StyleTransformData.cpp:
* Source/WebCore/rendering/svg/SVGTextLayoutEngineBaseline.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/animations/StyleAnimation.cpp:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.cpp:
* Source/WebCore/style/values/color/StyleDynamicRangeLimitMix.cpp:
* Source/WebCore/style/values/color/StyleRelativeColor.h:
* Source/WebCore/style/values/grid/StyleGridPosition.cpp:
* Source/WebCore/style/values/grid/StyleGridTemplateList.cpp:
* Source/WebCore/style/values/masking/StyleMaskBorder.cpp:
* Source/WebCore/style/values/masking/StyleMaskLayer.cpp:
* Source/WebCore/style/values/motion/StyleOffsetPosition.cpp:
* Source/WebCore/style/values/motion/StyleOffsetRotate.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+CSSValueConversion.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+CSSValueCreation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+Logging.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+Serialization.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+ValueRepresentationNeeded.h: Added.
* Source/WebCore/style/values/scroll-animations/StyleViewFunction.cpp:
* Source/WebCore/style/values/shapes/StylePathOperationWrappers.cpp:
* Source/WebCore/style/values/svg/StyleSVGGlyphOrientationHorizontal.cpp: Added.
* Source/WebCore/style/values/svg/StyleSVGGlyphOrientationHorizontal.h: Added.
* Source/WebCore/style/values/svg/StyleSVGGlyphOrientationVertical.cpp: Added.
* Source/WebCore/style/values/svg/StyleSVGGlyphOrientationVertical.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleMatrix3DTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleMatrixTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StylePerspectiveTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleRotateTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleScaleTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleSkewTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleTranslateTransformFunction.cpp:
* Source/WebCore/style/values/transitions/StyleTransition.cpp:

Canonical link: <a href="https://commits.webkit.org/301059@main">https://commits.webkit.org/301059@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/792342f9883e45d51d71a77ba4cf1f0c8dfdc130

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124838 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44510 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131683 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126715 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53079 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94997 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127792 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36071 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111650 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75559 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35005 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29806 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75160 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105830 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30035 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134350 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51683 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39491 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/103478 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52098 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107865 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103239 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26279 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48603 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26889 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48699 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51557 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/57365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50948 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54304 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52644 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/values/CSSValueConcepts.h:
```diff
@@ -69,6 +69,17 @@ template<typename> inline constexpr auto TreatAsVariantLike = false;
+// value representation" for CSS value type algorithms.
+// Requirements: Types must be an enum and specialize ValueRepresentation.
+// FIXME: This exists as a transitionary mechanism while not all enums specialize ValueRepresentation. Once they all do, this can be removed and std::is_enum_v<> can be used instead.
+template<typename> inline constexpr auto TreatAsEnumWithValueRepresentation = false;
```
Should we do this in the opposite direction, with a default of true and we specialize right out of the gate for all the cases that need to be false? Presumably if we forget one we will get a compilation failure. I’d love to have a to do list that goes with the FIXME above.

weinig replied:
Good idea!!

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveKeyword+ValueRepresentationNeeded.h:
```diff
@@ -0,0 +1,218 @@
+DEFINE_ENUM_NEEDS_VALUE_REPRESENTATION_CONFORMANCE(ContentVisibility)
+DEFINE_ENUM_NEEDS_VALUE_REPRESENTATION_CONFORMANCE(CSSBoxType)
+DEFINE_ENUM_NEEDS_VALUE_REPRESENTATION_CONFORMANCE(CursorType)
+DEFINE_ENUM_NEEDS_VALUE_REPRESENTATION_CONFORMANCE(CursorVisibility)
```
Looks like this needs to be wrapped in `#if ENABLE(CURSOR_VISIBILITY)`.

weinig committed:
[Style] Convert the 'glyph-orientation-*' properties to strong style types
https://bugs.webkit.org/show_bug.cgi?id=300064

Reviewed by Darin Adler.

Converts the 'glyph-orientation-horizontal' and 'glyph-orientation-vertical'
properties to strong style types.

To make properties like these, ones that can represented purely as `enum`,
work with strong style types, a new interface, `ValueRepresentation`, has
been added that allows converting the enum to its counterpart. This allows
us to implement a single interface and getting CSSValueCreation, Serialization
and Logging all for free. I expect this interface to evolve as we convert
more `enum` properties (for example, should we have a bidirectional mapping
of enum values <-> CSSValueIDs for simpler conversions, probably), but this
gives us a good starting out point.

To make it clear which enums still need `ValueRepresentation` conformance,
enums that don't have it (and are used with the style system) will cause
a compile error unless they are added to the list in StylePrimitiveKeyword+ValueRepresentationNeeded.h

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/ViewTimeline.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/CSSValueConcepts.h:
* Source/WebCore/page/IntersectionObserver.cpp:
* Source/WebCore/rendering/RenderTreeAsText.cpp:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/OutlineValue.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.cpp:
* Source/WebCore/rendering/style/StyleAppleColorFilterData.cpp:
* Source/WebCore/rendering/style/StyleBoxData.cpp:
* Source/WebCore/rendering/style/StyleDeprecatedFlexibleBoxData.cpp:
* Source/WebCore/rendering/style/StyleFilterData.cpp:
* Source/WebCore/rendering/style/StyleFlexibleBoxData.cpp:
* Source/WebCore/rendering/style/StyleFontData.cpp:
* Source/WebCore/rendering/style/StyleGridData.cpp:
* Source/WebCore/rendering/style/StyleInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMarqueeData.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMultiColData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleSurroundData.cpp:
* Source/WebCore/rendering/style/StyleTransformData.cpp:
* Source/WebCore/rendering/svg/SVGTextLayoutEngineBaseline.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/animations/StyleAnimation.cpp:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.cpp:
* Source/WebCore/style/values/color/StyleDynamicRangeLimitMix.cpp:
* Source/WebCore/style/values/color/StyleRelativeColor.h:
* Source/WebCore/style/values/grid/StyleGridPosition.cpp:
* Source/WebCore/style/values/grid/StyleGridTemplateList.cpp:
* Source/WebCore/style/values/masking/StyleMaskBorder.cpp:
* Source/WebCore/style/values/masking/StyleMaskLayer.cpp:
* Source/WebCore/style/values/motion/StyleOffsetPosition.cpp:
* Source/WebCore/style/values/motion/StyleOffsetRotate.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+CSSValueConversion.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+CSSValueCreation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+Logging.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+Serialization.h:
* Source/WebCore/style/values/primitives/StylePrimitiveKeyword+ValueRepresentationNeeded.h: Added.
* Source/WebCore/style/values/scroll-animations/StyleViewFunction.cpp:
* Source/WebCore/style/values/shapes/StylePathOperationWrappers.cpp:
* Source/WebCore/style/values/svg/StyleSVGGlyphOrientationHorizontal.cpp: Added.
* Source/WebCore/style/values/svg/StyleSVGGlyphOrientationHorizontal.h: Added.
* Source/WebCore/style/values/svg/StyleSVGGlyphOrientationVertical.cpp: Added.
* Source/WebCore/style/values/svg/StyleSVGGlyphOrientationVertical.h: Added.
* Source/WebCore/style/values/transforms/functions/StyleMatrix3DTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleMatrixTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StylePerspectiveTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleRotateTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleScaleTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleSkewTransformFunction.cpp:
* Source/WebCore/style/values/transforms/functions/StyleTranslateTransformFunction.cpp:
* Source/WebCore/style/values/transitions/StyleTransition.cpp:

Canonical link: https://commits.webkit.org/301059@main
